### ISSUE TITLE
dwi2fod msmt_csd : all zero output when nonassignable volume(s) to shell(s), with at least 2 tissues

### ISSUE THREAD

**Issue by thijsdhollander** [MRtrix3/mrtrix3#2129]
See title; encountered and described below the double line over here: https://community.mrtrix.org/t/dhollander-response-function-estimation/3895/3

**Comment by jdtournier** [comment]
I can't reproduce the issue you suggest with `dwi2fod` - though I can reproduce the `dwi2response` failure on Nagesh's data. Strangely enough, `dwi2response` runs to completion when invoked with the `-debug` option, and produces sensible looking responses (I've included them at the end). 

Also, I can run 2-tissue and 3-tissue CSD with these responses on Nagesh's data, and obtain sensible results as expected. I don't think the issue is with `dwi2fod`...

---

CFS response:
```
# Shells: 1,1000,2000,2999
# command_history: /home/donald/mrtrix3/bin/amp2response dwi.mif voxels_csf.mif safe_vecs.mif response_csf.txt -shells '2,1000,2000,3000' -isotropic -info  (version=3.0.1-dirty)
106728.822368227
5361.0960346865
1073.3319908781
1028.00530312388
```

GM response:
```
# Shells: 1,1000,2000,2999
# command_history: /home/donald/mrtrix3/bin/amp2response dwi.mif voxels_gm.mif safe_vecs.mif response_gm.txt -shells '2,1000,2000,3000' -isotropic -info  (version=3.0.1-dirty)
26010.5022240882
11845.6443789151
6179.09582232265
3620.06613681023
```

WM response:
```
# Shells: 1,1000,2000,2999
# command_history: /home/donald/mrtrix3/bin/amp2response dwi.mif voxels_sfwm.mif safe_vecs.mif response_sfwm.txt -shells '2,1000,2000,3000' -info  (version=3.0.1-dirty)
8985.65814467279 0 0 0 0 0
4695.93083470343 -1704.08797330743 351.980843983762 -19.0385718106989 -36.8681363286369 30.9737319511648
3269.0388671909 -1639.75189239719 677.371291295271 -173.512667931135 14.0403324921076 6.89306291481875
2656.78007487471 -1347.49375389639 711.137446874446 -279.933515873127 80.6283215235586 -24.0382346295036
```

**Comment by thijsdhollander** [comment]
The problem with Nagesh' data, before the ultimate crash, finds its origins at this line: https://github.com/MRtrix3/mrtrix3/blob/master/lib/mrtrix3/dwi2response/dhollander.py#L260 

To compute the SF WM metric, a 2-tissue MSMT-CSD is run; but for Nagesh' data, it returns all zero images, though with the correct dimensions (corresponding to an lmax=2 (6 volumes) and lmax=0 output for the 2 tissues in this particular case).  Down the track in the script, these then become NaN in the computation of the SF WM metric, and eventually `mrthreshold` fails on those.  But the problematic behaviour is at the `dwi2fod msmt_csd` run, where the output is all zero.

I didn't try to run it with `-debug`, but I did try to run a few variants of that line just to see whether it had anything to do with the particular responses used there or something like that.  It did produce (sensible) non-zero output when e.g. I ran it with only 1 tissue and 1 of the particular responses used there (i.e. `ewmrf.txt` or `response_csf.txt`).  But note that this particular line / run of MSMT-CSD also includes the `-mask` option, the `-lmax 2,0` option and the `-shells 2,1000,2000,3000` options, should these play a crucial role in triggering the behaviour.  I didn't check whether any of those are also crucial in reproducing the issue.  But running with 1 tissue with either response, I did still have all those options in place (only adjusting `-lmax` to only 2 or 0, depending on the response).  Hence the "with at least 2 tissues" I mentioned.

In any case, if it does work with the `-debug` option, I reckon you should also see non-zero outputs for that particular MSMT-CSD run, i.e. as they should be.
I did test by removing the 2 non-shell-assignable volumes up front as I suggested to Nagesh, and the observation is the same: that `dwi2fod msmt_csd` line returns the non-zero outputs as it should.

So when you reproduce the failure, you're likely also reproducing the particular `dwi2fod` behaviour, but under all specific conditions of that particular line / call of course.

**Comment by thijsdhollander** [comment]
Ok, some investigations later: it's incredibly specific.  I can rule out dependence on `-lmax` and `-shells`.  It has to do with the specific contents of the `response_csf.txt`, in combination with the non-fully-shell-assigned dataset.  And for some reason, it has to do with the exact contents of the comments at the start of that response function file.

**Comment by jdtournier** [comment]
That's odd... Any chance you can send me that problematic response file? Maybe there's some non-ascii or Unicode characters in that comment throwing off the `getline()` call?!?

**Comment by thijsdhollander** [comment]
Ok... narrowed it further down.  I tried all sorts of edits of the response file to indeed rule out if it's a specific character.  But instead, it seems to be related to the exact _length_ of the comments...?  What follows below would trigger the zero output for me at least, even when copy-pasting fresh from my post right here.  I've used the values from your WM and CSF responses above respectively, to further rule out it had anything to do with particular values or something.  Here goes:


Copy-paste this into a file called `a.txt` (no newline at the end, just the 4 lines; it's your WM response function values as above, but without comments):

```
8985.65814467279 0 0 0 0 0
4695.93083470343 -1704.08797330743 351.980843983762 -19.0385718106989 -36.8681363286369 30.9737319511648
3269.0388671909 -1639.75189239719 677.371291295271 -173.512667931135 14.0403324921076 6.89306291481875
2656.78007487471 -1347.49375389639 711.137446874446 -279.933515873127 80.6283215235586 -24.0382346295036
```


Copy-paste this into a file called `b.txt` (also no newline at the end, just the 6 lines):

```
# xxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
106728.822368227
5361.0960346865
1073.3319908781
1028.00530312388
```

This is your CSF response function values (as above), but I simplified the comments to rule out any particular special character's effect.  To reproduce what I'm seeing, the exact length, i.e. the number of `x` characters in these 2 comment lines, seems to be crucial.

Then, create some mask or ROI for CSD (not sure whether crucial, but I used a small manual drawn ROI to speed up testing; seems to work for different masks though).  I called it `roi.mif`.

Then run this, using Nagesh' original `dwi.mif` without modifications:

```
dwi2fod msmt_csd dwi.mif a.txt aa.mif b.txt bb.mif -mask roi.mif
```

Under all these conditions, `aa.mif` and `bb.mif` come out with only zero values (confirmed by `mrstats`).  If I remove some `x` characters in those 2 comment lines however, it suddenly works.  It depends on the number though; sometimes it works, sometimes it doesn't (but for a given setting and number of `x` characters, it seems to _consistently_ work, or not work).  It also works if I remove both comment lines entirely.  When the problem reproduces, the MSMT-CSD run finishes very quickly.  When the problem doesn't happen, it takes the normal amount of time (evident when using a larger `roi.mif` mask).

In the case where you can reproduce Nagesh' crash specifically in the context of `dwi2response`, if you go into the temp / scratch folder, you can also use the `response_csf.txt` instead of `b.txt`, which for me also reproduced the zero output.  I actually started with `response_csf.txt` in there, and gradually changed things to rule out special characters or specific numerical values, or even the filename.  In the end, I could get all the way to the example I provided here while still retaining the all zero output problem.

**Comment by jdtournier** [comment]
Wow, ok, that's unexpected. Thanks for drilling down into it. I'll have a look into what's going on there...

**Comment by jdtournier** [comment]
OK, that was nothing to do with the response files as such... The rows that corresponded to the excluded volumes weren't explicitly zeroed, so that meant sometimes it worked (the memory region had previously been zeroed by the OS on allocation), sometimes it didn't (the memory region was recycled from some previous use, and had not been explicitly zeroed). All this depends on what the memory allocator has allocated & deallocated so far, hence the dependence on a completely unrelated input - all it did was affect the size of a previous allocation...

Fix in #2132. :crossed_fingers:
